### PR TITLE
Use cTDS instead of pyODBC for more reliable SQL exports

### PR DESCRIPTION
**Pull request by evansd**
We have been experiencing problems downloading large cohorts over ODBC, both with pyODBC and Microsoft's own `sqlcmd` utility. This PR replaces pyODBC with [cTDS](https://pypi.org/project/ctds/) which uses the [Tabular Data Stream](https://en.wikipedia.org/wiki/Tabular_Data_Stream) protocol and doesn't seem to suffer from the same problem.

cTDS is a compiled extension and relies on the [FreeTDS](http://www.freetds.org/) library. To simplify installation we have packaged this up as a pre-built binary wheel at https://github.com/opensafely/ctds-binary. This is currently Linux only, so anyone wanting to run the cohortextractor on other platforms will need to install it themselves. However, most local uses of the cohortextractor (e.g. generating dummy data or updating codelists) don't require a database driver at all so we have made it possible to install and run the cohortextractor without one.

## Background

Attempting to download large (by which we mean more than about 2GB) results sets using pyODBC sometimes fails with the error:
```
pyodbc.OperationalError: (
  '08S01',
  '[08S01] [Microsoft][ODBC Driver 17 for SQL Server]TCP Provider: Error code 0x2746 (10054) (SQLGetData)'
)
```
These were somewhere intermittent and failed downloads would sometimes work on the next attempt. However the frequency seemed to increase with the size of the download to the point where sufficiently large results sets would fail every time.

Attempting to use [sqlcmd](https://docs.microsoft.com/en-us/sql/tools/sqlcmd-utility?view=sql-server-ver15) instead would fail in one of two ways. If we gave sqlcmd a filename to write its output to it would run without throwing an error, however it appeared to wrap its output at 2.1GB (i.e. the 32 bit address limit) and start writing later results over the beginning of the file.

This behaviour could be avoided by making it write to a pipe rather than a file. However after about 3GB of data this would fail with the error:
```
Sqlcmd: Error: Internal error at ReadAndHandleColumnData (Reason: Error reading column data).
TCP Provider: Error code 0x2746
Communication link failure
```

## TDS Protocol

Both pyODBC and sqlcmd use the ODBC protocol. However the same sets of results could be downloaded without problem using SQL Server Management Studio, which suggested that using some other protocol might be more reliable and a bit of searching brought up TDS (Tabular Data Stream). There are several Python packages which claim to speak this and offer a DB-API compatible interface to it. However, all but one had issues.

[python-tds](https://github.com/denisenkom/pytds) is a pure Python module which would make installation simple. However I couldn't get this to connect to the production database at all. Possibly this was something to do with not configuring TLS correctly (though python-tds does claim to support this.)

[pymssql](https://github.com/pymssql/pymssql) is, like cTDS, a wrapper around the FreeTDS library. It provides binary wheels for a range of platforms which, again, simplifies installation. It appears to be under active development, with a range of contributors, and was [praised](https://docs.sqlalchemy.org/en/13/dialects/mssql.html#module-sqlalchemy.dialects.mssql.pymssql) in the SQLAlchemy documentation. Initial testing seemed to go well, however when the set of results got large enough it would suddenly die with the error:
```
free(): corrupted unsorted chunks: 0x0000000001c49df0
``` 
This appears to be a [known bug](https://github.com/pymssql/pymssql/issues/500) related to garbage collection.

Finally we have [cTDS](https://github.com/zillow/ctds). This appears to be a bit of a one man show, but it's under the account of a major web company and it looks like a long-lived and consistently maintained project so hopefully they're using it in production and have an incentive to keep it working. In any case, if we start of having issues with this we can revisit the other two libraries. The main issue here is that there are no binary wheels, but building for just the platform we need turned out not to be too difficult.

### CODE REVIEW

File: cohortextractor/mssql_utils.py [hunk 1]
```diff
@@ -21,107 +25,59 @@ def mssql_connection_params_from_url(url):
...
 
-def mssql_pyodbc_connection_from_url(url):
+def mssql_dbapi_connection_from_url(url):
     # We avoid importing this immediately because we're using the TPPBackend
     # code for error checking (as a hopefully temporary shortcut) and so want
-    # to be able to create a TPPBackend instance without needing pyodbc
-    # installed (which complicates local installing).
-    import pyodbc
+    # to be able to create a TPPBackend instance without needing a MSSQL
+    # database driver installed (which complicates local installing).
```
Comment (sebbacon) [inline cohortextractor/mssql_utils.py:32]: Perhaps adding a "why ctds?" note which links to this PR description here?

**Review by sebbacon — APPROVED** [review]
:heart_decoration: 

I wonder if we should remove the temporary database stuff? We added it because of odd errors which we've reason to think will now be resolved; there is a theoretical debugging benefit but only if we use it by default (I think) and it needs more tooling to help identify the name of the temporary databases used in previous runs.

**Comment by evansd** [comment]
Yes, hopefully this will put all our connection woes behind us and we can get rid of all the temporary database stuff. I think we should leave it in place for now though as it was absolutely invaluable in debugging this issue and I don't want to tempt fate by getting rid of it just yet!

File: requirements.in [hunk 2]
```diff
@@ -1,11 +1,25 @@
...
+# If you're on a platform other than Linux then you'll have to install cTDS yourself:
+# https://zillow.github.io/ctds/install.html
+# https://pypi.org/project/ctds/
+#
+https://github.com/opensafely/ctds-binary/raw/b76175dd04baace2327518a78536d35bac6c7e2b/dist/ctds-1.12.0-cp37-cp37m-manylinux2014_x86_64.whl ; sys_platform == "linux"
+
+# We still use pyODBC in our tests because we use SQLAlchemy to set up our test
+# data and there's currently no SQLAlchemy adapter for cTDS.  I managed to get
+# one mostly working, but it tripped up on the hyphen separated field names in
+# the EMIS backend. Once these have been removed it might be worth trying
```
Comment (inglesp) [inline requirements.in:26]: I've just merged #280, so the hyphens are gone.

**Review by inglesp — APPROVED** [review]
This is great, well done.

File: requirements.in (hunk 2 above)
Comment (evansd) [reply requirements.in:24]: Thanks. I've stuck my WIP in a [branch](https://github.com/opensafely/cohort-extractor/tree/remove-pyodbc) so we can come back to it later.

**Comment by evansd** [comment]
Further notes in case we have to revisit this in future. There's a [mssql-cli](https://github.com/dbcli/mssql-cli) tool which looks to be an [officially supported](https://docs.microsoft.com/en-us/sql/tools/mssql-cli?view=sql-server-ver15) Microsoft thing and an improved successor to `sqlcmd`.

As it happens this is a pip installable Python tool (with pre-built wheels). It uses JSON-RPC to talk to SQL Server which gives us yet another protocol to try in case cTDS gives us problems. It doesn't currently support CSV output, though this is apparently [being worked on](https://github.com/dbcli/mssql-cli/issues/105) and given that it's all Python we could possibly hack this in ourselves if we had to.

**Pull request merged**